homeassistant/components/mqtt/__init__.py

```diff
@@ -585,6 +584,51 @@ class Message:
 class MQTT:
     """Home Assistant MQTT client."""
 
+    class AsyncioHelper:
+        """Helper to handle paho-mqtt async callbacks."""
+
+        def __init__(self, hass, client):
+            """Initialize helper."""
+            self.hass = hass
+            self.task_misc = None
+            self.client = client
+            self.client.on_socket_open = self.on_socket_open
+            self.client.on_socket_close = self.on_socket_close
+            self.client.on_socket_register_write = self.on_socket_reg_write
+            self.client.on_socket_unregister_write = self.on_socket_unreg_write
+
+        def on_socket_open(self, client, userdata, sock):
+            """Socket open callback."""
+            def cb():
+                client.loop_read()
+
+            self.hass.loop.add_reader(sock, cb)
+            self.task_misc = self.hass.async_create_task(self.misc_loop())
+
+        def on_socket_close(self, client, userdata, sock):
+            """Socket closed callback."""
+            self.hass.loop.remove_reader(sock)
+            self.task_misc.cancel()
+
+        def on_socket_reg_write(self, client, userdata, sock):
+            """Schedule socket write callback."""
+            def cb():
+                client.loop_write()
+
+            self.hass.loop.add_writer(sock, cb)
+
+        def on_socket_unreg_write(self, client, userdata, sock):
+            """Unschedule socket write callback."""
+            self.hass.loop.remove_writer(sock)
+
+        async def misc_loop(self):
+            """Peridocially call loop_misc()."""
+            while self.client.loop_misc() == 0:
```

Comment:
In the paho example they use a constant from the library for the check. Maybe it doesn't matter?

Reply:
I aligned with the existing code which also doesn't use the constant. If you think it's cleaner to use the constant, I'll do so?

Reply:
They could change the constant, and then we could break. That's my worry.

Reply:
Alright, replaced all comparisons with 0 to compare with MQTT_ERR_SUCCESS